File: src/app/ports/persistence_port.ts
import type { BotConfig, Pair, RunRecord, TradeRecord } from '../../domain/model/types';

export interface PersistencePort {
  getCurrentConfig(): Promise<BotConfig>;
  createTrade(trade: TradeRecord): Promise<void>;
  updateTrade(tradeId: string, updates: Partial<TradeRecord>): Promise<void>;
  findOpenTrade(pair: Pair): Promise<TradeRecord | null>;
  countTradesForUtcDay(pair: Pair, dayStartIso: string, dayEndIso: string): Promise<number>;
  saveRun(run: RunRecord): Promise<void>;
}

File: pybot/app/ports/persistence_port.py
from __future__ import annotations

from typing import Protocol

from pybot.domain.model.types import BotConfig, Pair, RunRecord, TradeRecord


class PersistencePort(Protocol):
    def get_current_config(self) -> BotConfig: ...

    def create_trade(self, trade: TradeRecord) -> None: ...

    def update_trade(self, trade_id: str, updates: dict) -> None: ...

    def find_open_trade(self, pair: Pair) -> TradeRecord | None: ...

    def count_trades_for_utc_day(self, pair: Pair, day_start_iso: str, day_end_iso: str) -> int: ...

    def save_run(self, run: RunRecord) -> None: ...

